File: sdk-go/pagination/auto_paginator.go
// Package pagination contains the generic auto-paginating iterator returned
// by every list endpoint's auto-paginate variant.
package pagination

// Iter generic auto-paginating iterator returned by every list endpoint's
// auto-paginate variant. It walks pages lazily — one HTTP call per page, only
// when the consumer drains the previous page's buffer.
//
// Idiomatic usage:
//
//	iter := api.Events.ListAutoPaginate(ctx, nil)
//	for iter.Next() {
//		ev := iter.Current()
//		_ = ev
//	}
//	if err := iter.Err(); err != nil {
//		// non-nil if any page failed
//	}
//
// On Go 1.23+ a range-over-func variant is also supported — see [Iter.All].
type Iter[T any] struct {
	fetchPage func(page int) ([]T, bool, error)
	page      int
	buffer    []T
	index     int
	hasMore   bool
	current   T
	err       error
}

// NewIter constructs an [*Iter] backed by fetchPage. fetchPage receives a
// 0-indexed page number and returns (data, hasMore, err). Not part of the
// public API; resource packages call this from their list-auto-paginate
// methods.
//
//nolint:revive // referenced from resource packages
func NewIter[T any](startPage int, fetchPage func(page int) ([]T, bool, error)) *Iter[T] {
	return &Iter[T]{
		fetchPage: fetchPage,
		page:      startPage,
		hasMore:   true,
	}
}

// Next advances the iterator. Returns true on success, false when the stream
// is exhausted or an error occurred. After Next returns false, callers should
// check [Iter.Err] to distinguish end-of-stream from failure.
func (it *Iter[T]) Next() bool {
	if it.err != nil {
		return false
	}

	if it.index < len(it.buffer) {
		it.current = it.buffer[it.index]
		it.index++
		return true
	}

	if !it.hasMore {
		return false
	}

	data, hasMore, err := it.fetchPage(it.page)
	if err != nil {
		it.err = err
		return false
	}
	it.buffer = data
	it.index = 0
	it.hasMore = hasMore
	it.page++

	if len(it.buffer) == 0 {
		return false
	}

	it.current = it.buffer[it.index]
	it.index++
	return true
}

// Current returns the most recent value yielded by [Iter.Next]. Calling
// Current before the first successful Next, or after Next has returned false,
// returns the zero value of T.
func (it *Iter[T]) Current() T { return it.current }

// Err returns the first error encountered during iteration, or nil if
// iteration completed cleanly.
func (it *Iter[T]) Err() error { return it.err }

File: sdk-python/src/threecommon/pagination/auto_paginator.py
"""auto-paginating iterators returned by every list endpoint.

Two flavors:

* [Iter][threecommon.pagination.Iter] — sync iterator. Use ``for ev in iter:``.
* [AsyncIter][threecommon.pagination.AsyncIter] — async iterator.
  Use ``async for ev in iter:``.

Both walk pages lazily — one HTTP call per page, only when the user
drains the previous page's buffer.
"""

from __future__ import annotations

from collections.abc import AsyncIterator, Awaitable, Callable, Iterator
from typing import Generic, TypeVar

T = TypeVar("T")

#: Synchronous page-fetch callback. Returns ``(page_items, has_more)``.
SyncFetchPage = Callable[[int], tuple[list[T], bool]]

#: Asynchronous page-fetch callback. Returns ``(page_items, has_more)``.
AsyncFetchPage = Callable[[int], Awaitable[tuple[list[T], bool]]]


class Iter(Generic[T]):
    """Sync auto-paginating iterator. ``for ev in iter:`` drives the pages."""

    __slots__ = ("_buffer", "_fetch_page", "_has_more", "_index", "_page")

    def __init__(self, *, fetch_page: SyncFetchPage[T], start_page: int = 0) -> None:
        self._fetch_page = fetch_page
        self._page = start_page
        self._buffer: list[T] = []
        self._index = 0
        self._has_more = True

    def __iter__(self) -> Iterator[T]:
        return self

    def __next__(self) -> T:
        if self._index < len(self._buffer):
            value = self._buffer[self._index]
            self._index += 1
            return value
        if not self._has_more:
            raise StopIteration

        data, has_more = self._fetch_page(self._page)
        self._buffer = data
        self._index = 0
        self._has_more = has_more
        self._page += 1

        if not self._buffer:
            raise StopIteration

        value = self._buffer[self._index]
        self._index += 1
        return value


class AsyncIter(Generic[T]):
    """Async auto-paginating iterator. ``async for ev in iter:`` drives the pages."""

    __slots__ = ("_buffer", "_fetch_page", "_has_more", "_index", "_page")

    def __init__(self, *, fetch_page: AsyncFetchPage[T], start_page: int = 0) -> None:
        self._fetch_page = fetch_page
        self._page = start_page
        self._buffer: list[T] = []
        self._index = 0
        self._has_more = True

    def __aiter__(self) -> AsyncIterator[T]:
        return self

    async def __anext__(self) -> T:
        if self._index < len(self._buffer):
            value = self._buffer[self._index]
            self._index += 1
            return value
        if not self._has_more:
            raise StopAsyncIteration

        data, has_more = await self._fetch_page(self._page)
        self._buffer = data
        self._index = 0
        self._has_more = has_more
        self._page += 1

        if not self._buffer:
            raise StopAsyncIteration

        value = self._buffer[self._index]
        self._index += 1
        return value
